src/impl_methods.rs

```diff
@@ -3184,6 +3184,90 @@ impl<A, D: Dimension> ArrayRef<A, D>
             f(&*prev, &mut *curr)
         });
     }
+
+    /// Return a partitioned copy of the array.
+    ///
+    /// Creates a copy of the array and partially sorts it around the k-th element along the given axis.
+    /// The k-th element will be in its sorted position, with:
+    /// - All elements smaller than the k-th element to its left
+    /// - All elements equal or greater than the k-th element to its right
+    /// - The ordering within each partition is undefined
+    ///
+    /// # Parameters
+    ///
+    /// * `kth` - Index to partition by. The k-th element will be in its sorted position.
+    /// * `axis` - Axis along which to partition.
+    ///
+    /// # Returns
+    ///
+    /// A new array of the same shape and type as the input array, with elements partitioned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ndarray::prelude::*;
+    ///
+    /// let a = array![7, 1, 5, 2, 6, 0, 3, 4];
+    /// let p = a.partition(3, Axis(0));
+    ///
+    /// // The element at position 3 is now 3, with smaller elements to the left
+    /// // and greater elements to the right
+    /// assert_eq!(p[3], 3);
+    /// assert!(p.slice(s![..3]).iter().all(|&x| x <= 3));
+    /// assert!(p.slice(s![4..]).iter().all(|&x| x >= 3));
+    /// ```
+    pub fn partition(&self, kth: usize, axis: Axis) -> Array<A, D>
+    where A: Clone + Ord
+    {
+        // Check if axis is valid
+        if axis.index() >= self.ndim() {
+            panic!("axis {} is out of bounds for array of dimension {}", axis.index(), self.ndim());
+        }
+
+        // Check if kth is valid
+        if kth >= self.len_of(axis) {
+            panic!("kth {} is out of bounds for axis {} with length {}", kth, axis.index(), self.len_of(axis));
+        }
+
+        // If the array is empty, return a copy
+        if self.is_empty() {
+            return self.to_owned();
+        }
+
+        // If the array is 1D, handle as a special case
+        if self.ndim() == 1 {
+            let mut result = self.to_owned();
+            if let Some(slice) = result.as_slice_mut() {
+                slice.select_nth_unstable(kth);
+            }
+            return result;
+        }
+
+        // For multi-dimensional arrays, partition along the specified axis
+        let mut result = self.to_owned();
+
+        // Process each lane with partitioning
+        Zip::from(result.lanes_mut(axis)).for_each(|mut lane| {
+            // For each lane, we need a contiguous slice to partition
+            if let Some(slice) = lane.as_slice_mut() {
+                // If the lane's memory is contiguous, use select_nth_unstable directly
+                slice.select_nth_unstable(kth);
+            } else {
+                // For non-contiguous memory, create a temporary vector
+                let mut values = lane.iter().cloned().collect::<Vec<_>>();
```

Comment:
A single vector should be declared, not N vectors. I'm pretty sure it's possible to know in advance (before the `Zip`) if the lanes will be contiguous or not. I **think** we can safely say that if the first lane is contiguous, then all other lanes will be too.